eidas_proxy_service/datamodels.py

```diff
@@ -0,0 +1,200 @@
+                if required:
+                    raise ValidationError(
+                        {name: 'Must be {}, not {}.'.format(required_type.__name__, type(value).__name__)})
+                if value is not None:
+                    raise ValidationError(
+                        {name: 'Must be {} or None, not {}.'.format(required_type.__name__, type(value).__name__)})
+
+    def __iter__(self) -> Iterator[Any]:
+        """Iterate over values of all fields."""
+        return (getattr(self, name) for name in self.FIELDS)
+
+    def __repr__(self) -> str:
+        return '{}({})'.format(
+            self.__class__.__name__,
+            ', '.join('{}={!r}'.format(key, getattr(self, key)) for key in self.FIELDS)
+        )
+
+    def __str__(self) -> str:
+        return repr(self)
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, DataModel):
+            return NotImplemented
+        return type(self) is type(other) and self.get_data_as_tuple() == other.get_data_as_tuple()
+
+
+T = TypeVar('T', bound='XMLDataModel')
+
+
+class XMLDataModel(DataModel, ABC):
+    """Data model with XML serialization and deserialization."""
+
+    ROOT_ELEMENT = None  # type: str
+    """The name of the root element."""
+
+    def export_xml(self) -> Element:
+        """
+        Export LightRequest as a XML document.
+
+        :return: A XML document.
+        :raise ValidationError: If the model validation fails.
+        """
+        self.validate()
+        if not self.ROOT_ELEMENT:
+            raise TypeError('XMLDataModel subclasses must define ROOT_ELEMENT class attribute.')
+        root = Element(self.ROOT_ELEMENT)
+        self.serialize_fields(root)
+        return root
+
+    def serialize_fields(self, parent_element: Element) -> None:
+        """Serialize model fields."""
+        for field_name in self.FIELDS:
+            value = getattr(self, field_name)
+            tag = convert_field_name_to_tag_name(field_name)
+            serialize_func = getattr(self, 'serialize_' + field_name, None)
+            if serialize_func:
+                serialize_func(parent_element, tag, value)
+            elif value is not None:
+                if isinstance(value, XMLDataModel):
+                    value.serialize_fields(SubElement(parent_element, tag))
+                else:
+                    if isinstance(value, Enum):
+                        value = value.value
+                    elif isinstance(value, bool):
+                        value = str(value).lower()
+                    else:
+                        value = str(value)
+                    SubElement(parent_element, tag).text = value
+
+    @classmethod
+    def load_xml(cls: Type[T], root: Union[Element, ElementTree]) -> T:
+        """
+        Load Light Request from a XML document.
+
+        :param root: The XML document to load.
+        :raise ValueError: If the XML document does have a valid Light Request structure.
+        """
```

Comment:
Models are not loaded from an XML string directly because we will sometimes need to [fix an invalid response](https://trac.nic.cz/ticket/23408#comment:17) first.